builtin/game/item_entity.lua

```diff
@@ -207,7 +207,12 @@ core.register_entity(":__builtin:item", {
 		end
 	end,
 
-	on_punch = function(self, hitter)
+	on_punch = function(self, hitter, ...)
+		local item_def = core.registered_items[self.itemstring]
```

Comment:
if you want to do this in Lua, all this added block should be in the next condition (if inv and self.itemstring != "")